numba/np/linalg.py

```diff
@@ -2792,9 +2792,10 @@ def kron_impl(a, b):
     fix_b = _kron_normaliser_impl(b)
     ret_c = _kron_return(a, b)
 
-    # this is fine because the ufunc for the Hadamard product
-    # will reject differing dtypes in a and b.
     dt = getattr(a, 'dtype', a)
+    if (isinstance(a, types.Array) and isinstance(b, types.Array) and
+       a.dtype != b.dtype):
+        raise NumbaTypeError("Both types must be the same in numpy.kron.")
```

Comment:
Do the `dtype`s need to be _the same_ or just of the same family, i.e. does this implementation work ok for two different floating point types?

Reply:
I have no idea. I'm matching the comment as best I understand it.

Reply:
The issue is here, the `dtype` is taken from `a`, the first arg:
https://github.com/numba/numba/blob/fb952f8f5959ce31583da612a91b204bca58da81/numba/np/linalg.py#L2797
which is then used as the `dtype` in the allocation of the return value (array `c`):
https://github.com/numba/numba/blob/fb952f8f5959ce31583da612a91b204bca58da81/numba/np/linalg.py#L2812-L2813
It is safest to require them to be the same, anything else would probably require analysis similar to that needed to just fix the issue!